File: codes/breadth_first_search.js
// Tree data structure
var TreeNode = function(node, left=null, right=null) {
	this.node = node;
	this.left = left;
	this.right = right;
};


// function to get children objects of a node
TreeNode.prototype.getChildren = function() {
    var children = [];
	if (this.left !== null) {
		children.push(this.left);
	}
	if (this.right !== null) {
		children.push(this.right);
	}
	return children;
};


// function to build a tree
var make_tree = function(index=0) {
	// Full binary tree (#7-nodes)
	if (index === 1) {
		console.log("Rendering full binary tree with 7 nodes...")
		t1 = new TreeNode(4, new TreeNode(6), new TreeNode(7));
		t2 = new TreeNode(2, t1, new TreeNode(5));
		t = new TreeNode(1, t2, new TreeNode(3));
		return t;
	} 
	
	// Complete binary tree (#10-nodes)
	else if (index === 2) {
		console.log("Rendering complete binary tree with 10 nodes...")
		t1 = new TreeNode(4, new TreeNode(8), new TreeNode(9));
		t2 = new TreeNode(2, t1, new TreeNode(5, new TreeNode(0))); 
		t3 = new TreeNode(3, new TreeNode(6), new TreeNode(7));
		t = new TreeNode(1, t2, t3);
		return t;
	} 
	// Perfect binary tree (#15-nodes)
	else {	
		if (index != 0) {
			console.log("Index not defined. Used 0 by default...")
		}
		console.log("Rendering perfect binary tree with 15 nodes...")
		t1 = new TreeNode(4, new TreeNode(8), new TreeNode(9))
		t2 = new TreeNode(5, new TreeNode(0), new TreeNode(1))
		t3 = new TreeNode(6, new TreeNode(2), new TreeNode(3))
		t4 = new TreeNode(7, new TreeNode(4), new TreeNode(5))
		t5 = new TreeNode(2, t1, t2)
		t6 = new TreeNode(3, t3, t4)
		t = new TreeNode(1, t5, t6)
		return t
	}
};


// function for the Inorder Traversal of binary tree
var inorder_traversal = function(tree, inorder_path) {
	if (tree != null) {
		inorder_traversal(tree.left, inorder_path);
		inorder_path.push(tree.node);
		inorder_traversal(tree.right, inorder_path);
	}
	return inorder_path;
};


// function for the binary search tree
var breadth_first_search = function(root) {
	var queue = [root];
	var bfs_result = [];
	
	while (queue.length != 0) {
		current = queue.shift();
		bfs_result.push(current.node);
		
		if (current.left != null) {
			queue.push(current.left)
		}
		if (current.right != null) {
			queue.push(current.right)
		}
	}
	return bfs_result;
};


// main function
function main() {
	var tree = make_tree(1);
	// console.log(tree.getChildren());
	var inorder_path = [];
	inorder_traversal(tree, inorder_path);
	console.log("Inorder Traversal > " + inorder_path);
	
	var bfs_path = breadth_first_search(tree);
	console.log("Breadth First Search > " + bfs_path);
}


main()

File: codes/breadth_first_search.go
package main;

import (
	"fmt"
	// "reflect"
	// "container/list"
)

// Tree data structure
type Node struct {
	node int
	left *Node
	right *Node
}

func (n Node) getNode() int {
	return n.node
}

func (n Node) getChildren() []*Node {
	_children := []*Node {}
	if n.left != nil {
		_children = append(_children, n.left)
	}
	if n.right != nil {
		_children = append(_children, n.right)
	}
	return _children;
}


func build_tree(args ...int) Node {
	if len(args) == 1 && args[0] == 1 {
		/*
		Full binary tree (#7 nodes)

				 1
			   ╱   ╲
		      2     3
		    ╱  ╲
		   4     5
		 ╱  ╲
		6    7
		*/
		
		fmt.Println("Rendering full binary tree with 7 nodes...")
		t1 := &Node{4, &Node{6, nil, nil}, &Node{7, nil, nil}}
		t2 := &Node{2, t1, &Node{5, nil, nil}}
		t := Node{1, t2, &Node{3, nil, nil}}
		return t
	} else if len(args) == 1 && args[0] == 2{
		/*
		Complete binary tree (#10 nodes)
			      1
			    ╱   ╲
		      2       3
		    ╱   ╲    ╱ ╲
		   4     5  6   7
		 ╱  ╲	╱
		8    9 0
		*/
		
		fmt.Println("Rendering complete binary tree with 10 nodes...")
		t1 := &Node{4, &Node{8, nil, nil}, &Node{9, nil, nil}}
		t2 := &Node{2, t1, &Node{5, &Node{0, nil, nil}, nil}}
		t3 := &Node{3, &Node{6, nil, nil}, &Node{7, nil, nil}}
		t := Node{1, t2, t3}
		return t
	} else {
		if len(args) == 1 && args[0] != 0{
			println("Invalid Index. Used 0 by default...")
		}
		/*
		Perfect binary tree (#15 nodes)
			         1
			   	  ╱     ╲
			   	 ╱       ╲
		       2           3
		    ╱   ╲        ╱   ╲
		  4      5      6	  7
		 ╱ ╲	╱ ╲    ╱ ╲   ╱  ╲
		8   9  0   1  2   3 4    5
		*/
		
		fmt.Println("Rendering complete binary tree with 15 nodes...")
		t1 := &Node{4, &Node{8, nil, nil}, &Node{9, nil, nil}}
		t2 := &Node{5, &Node{0, nil, nil}, &Node{1, nil, nil}}
		t3 := &Node{6, &Node{2, nil, nil}, &Node{3, nil, nil}}
		t4 := &Node{7, &Node{4, nil, nil}, &Node{5, nil, nil}}
		t5 := &Node{2, t1, t2}
		t6 := &Node{3, t3, t4}
		t := Node{1, t5, t6}
		return t
	}
}

func inorder_traversal(t *Node, _path *[]int){
	if (t != nil) {
		inorder_traversal(t.left, _path)
		*_path = append(*_path, t.node)
		inorder_traversal(t.right, _path)
	}
}

func breadth_first_search(root *Node) []int{
	queue := []*Node {root}
	var _path []int

	for len(queue) != 0 {
		current := queue[0]
		queue = queue[1:]
		_path = append(_path, current.node)
		for _, child := range current.getChildren() {
			queue = append(queue, child)
		}
	}
	return _path
}

func main() {
	tree := build_tree(1)
	fmt.Println("Tree >", tree)
	fmt.Println("Tree children >> ", tree.getChildren())

	inorder_path := []int{}
	inorder_traversal(&tree, &inorder_path)
	fmt.Println("Inorder traversal > ", inorder_path)

	bfs_path := breadth_first_search(&tree)
	fmt.Println("Breadth First Search > ", bfs_path)
}
